Python/ceval.c

```diff
@@ -881,6 +881,226 @@ _Py_CheckRecursiveCall(PyThreadState *tstate, const char *where)
     return 0;
 }
 
+
+// PEP 634: Structural Pattern Matching
+
+
+static PyObject*
+match_keys(PyThreadState *tstate, PyObject *map, PyObject *keys)
+{
+    // Return a tuple of values corresponding to keys, with error checks for
+    // duplicate/missing keys.
+    assert(PyTuple_CheckExact(keys));
+    Py_ssize_t nkeys = PyTuple_GET_SIZE(keys);
+    if (!nkeys) {
+        // No keys means no items.
+        return PyTuple_New(0);
+    }
+    PyObject *seen = NULL;
+    PyObject *dummy = NULL;
+    PyObject *values = NULL;
+    // We use the two argument form of map.get(key, default) for two reasons:
+    // - Atomically check for a key and get its value without error handling.
+    // - Don't cause key creation or resizing in dict subclasses like
+    //   collections.defaultdict that define __missing__ (or similar).
+    _Py_IDENTIFIER(get);
+    PyObject *get = _PyObject_GetAttrId(map, &PyId_get);
+    if (get == NULL) {
+        goto fail;
+    }
+    seen = PySet_New(NULL);
+    if (seen == NULL) {
+        goto fail;
+    }
+    // dummy = object()
+    dummy = _PyObject_CallNoArg((PyObject *)&PyBaseObject_Type);
+    if (dummy == NULL) {
+        goto fail;
+    }
+    values = PyTuple_New(nkeys);
```

Comment:
This is dangerous. You are preparing a tuple with `nkeys` but then you are filling it calling `PyObject_CallFunctionObjArgs`, which can execute arbitrary Python code. At this point, the tuple is tracked by the GC so code executed by `PyObject_CallFunctionObjArgs` could see the tuple half-initialized (the same if there is a gc pass).

Reply:
I see. So is filling it with `None` the right thing to do here, or can we cheat and just untrack/retrack?

Reply:
Actually, filling it with `None` can lead to problems (as we've seen before). So I'm guessing we should just untrack it temporarily?

Reply:
Okay, I just went with list -> tuple for both. Let me know if you prefer something else.

Reply:
Another option it to untrack the tuple after its creation and track it back when is ready. Be careful because some APIs can retrack it (like resizing the tuple). I would recommend to add some asserts to check that is untracked where is supposed to be if you decide to go this way.